File: ios/Gomoku/GomokuAI.swift
import Foundation

class GomokuAI {
    // Scoring constants
    static let FIVE: Int = 1_000_000
    static let LIVE_FOUR: Int = 50_000
    static let RUSH_FOUR: Int = 5_000
    static let LIVE_THREE: Int = 5_000
    static let SLEEP_THREE: Int = 500
    static let LIVE_TWO: Int = 500
    static let SLEEP_TWO: Int = 50

    let depth: Int
    let candidateRange: Int
    let maxCandidates: Int
    let randomPick: Bool

    init(depth: Int = 4, candidateRange: Int = 1, maxCandidates: Int = 10, randomPick: Bool = false) {
        self.depth = depth
        self.candidateRange = candidateRange
        self.maxCandidates = maxCandidates
        self.randomPick = randomPick
    }

    func findBestMove(board: [[Int]], aiPlayer: Int = GameBoard.black) -> (row: Int, col: Int)? {
        let candidates = getCandidates(board: board)
        guard !candidates.isEmpty else { return nil }

        var bestScore = Int.min
        var bestMoves: [(row: Int, col: Int)] = []
        let maximizingPlayer = aiPlayer

        let currentPlayer = getCurrentPlayer(board: board)

        for move in candidates {
            var newBoard = board.map { $0 }
            newBoard[move.row][move.col] = currentPlayer
            let score = minimax(board: newBoard, depth: depth - 1, alpha: Int.min, beta: Int.max,
                               maximizingPlayer: maximizingPlayer, isAITurn: false)

            if score > bestScore {
                bestScore = score
                bestMoves = [move]
            } else if score == bestScore {
                bestMoves.append(move)
            }
        }

        if randomPick && bestMoves.count > 1 {
            return bestMoves.randomElement()
        }
        return bestMoves.first
    }

    func getCandidates(board: [[Int]]) -> [(row: Int, col: Int)] {
        let hasMoves = board.contains { row in row.contains { $0 != GameBoard.empty } }
        if !hasMoves {
            return [(row: GameBoard.size / 2, col: GameBoard.size / 2)]
        }

        return selectTopCandidates(board: board)
    }

    // MARK: - Candidate Selection

    func selectTopCandidates(board: [[Int]]) -> [(row: Int, col: Int)] {
        let size = GameBoard.size
        var scored: [(row: Int, col: Int, score: Int)] = []

        // Check if board is empty or nearly empty for center placement
        var hasPieces = false
        for r in 0..<size {
            for c in 0..<size {
                if board[r][c] != GameBoard.empty {
                    hasPieces = true
                    break
                }
            }
            if hasPieces { break }
        }

        if !hasPieces {
            return [(row: size / 2, col: size / 2)]
        }

        // Build occupied set for neighbor checking
        var occupied = [(Int, Int)]()
        for r in 0..<size {
            for c in 0..<size {
                if board[r][c] != GameBoard.empty {
                    occupied.append((r, c))
                }
            }
        }

        // For each empty cell near occupied cells, compute a quick heuristic score
        var visited = Set<Int>()
        for (r, c) in occupied {
            for dr in -candidateRange...candidateRange {
                for dc in -candidateRange...candidateRange {
                    let nr = r + dr
                    let nc = c + dc
                    guard nr >= 0, nr < size, nc >= 0, nc < size else { continue }
                    guard board[nr][nc] == GameBoard.empty else { continue }
                    let key = nr * size + nc
                    guard !visited.contains(key) else { continue }
                    visited.insert(key)

                    let score = quickHeuristic(board: board, row: nr, col: nc)
                    scored.append((row: nr, col: nc, score: score))
                }
            }
        }

        scored.sort { $0.score > $1.score }
        let count = min(maxCandidates, scored.count)
        return Array(scored.prefix(count)).map { (row: $0.row, col: $0.col) }
    }

    func quickHeuristic(board: [[Int]], row: Int, col: Int) -> Int {
        var score = 0
        let directions = [(0, 1), (1, 0), (1, 1), (1, -1)]
        let size = GameBoard.size

        for (dr, dc) in directions {
            for player in [GameBoard.black, GameBoard.white] {
                var count = 0
                var blocked = 0

                // Forward
                for i in 1...4 {
                    let r = row + dr * i
                    let c = col + dc * i
                    guard r >= 0, r < size, c >= 0, c < size else { blocked += 1; break }
                    if board[r][c] == player {
                        count += 1
                    } else if board[r][c] != GameBoard.empty {
                        blocked += 1; break
                    } else { break }
                }

                // Backward
                for i in 1...4 {
                    let r = row - dr * i
                    let c = col - dc * i
                    guard r >= 0, r < size, c >= 0, c < size else { blocked += 1; break }
                    if board[r][c] == player {
                        count += 1
                    } else if board[r][c] != GameBoard.empty {
                        blocked += 1; break
                    } else { break }
                }

                let patternScore = patternScore(count: count, blocked: blocked)
                if player == GameBoard.black {
                    score += patternScore
                } else {
                    score += patternScore / 2
                }
            }
        }

        return score
    }

    // MARK: - Minimax with Alpha-Beta

    private func minimax(board: [[Int]], depth: Int, alpha: Int, beta: Int,
                         maximizingPlayer: Int, isAITurn: Bool) -> Int {
        if depth == 0 {
            return evaluateBoard(board: board, aiPlayer: maximizingPlayer)
        }

        let candidates = getCandidates(board: board)
        if candidates.isEmpty {
            return evaluateBoard(board: board, aiPlayer: maximizingPlayer)
        }

        let currentPlayer = getCurrentPlayer(board: board)

        if isAITurn {
            var value = Int.min
            var a = alpha
            for move in candidates {
                var newBoard = board.map { $0 }
                newBoard[move.row][move.col] = currentPlayer
                if checkWinOnBoard(newBoard, row: move.row, col: move.col, player: currentPlayer) {
                    return GomokuAI.FIVE * 10
                }
                value = max(value, minimax(board: newBoard, depth: depth - 1,
                                          alpha: a, beta: beta,
                                          maximizingPlayer: maximizingPlayer, isAITurn: false))
                a = max(a, value)
                if a >= beta { break }
            }
            return value
        } else {
            var value = Int.max
            var b = beta
            for move in candidates {
                var newBoard = board.map { $0 }
                newBoard[move.row][move.col] = currentPlayer
                if checkWinOnBoard(newBoard, row: move.row, col: move.col, player: currentPlayer) {
                    return -GomokuAI.FIVE * 10
                }
                value = min(value, minimax(board: newBoard, depth: depth - 1,
                                          alpha: alpha, beta: b,
                                          maximizingPlayer: maximizingPlayer, isAITurn: true))
                b = min(b, value)
                if alpha >= b { break }
            }
            return value
        }
    }

    private func getCurrentPlayer(board: [[Int]]) -> Int {
        var blackCount = 0
        var whiteCount = 0
        for row in board {
            for cell in row {
                if cell == GameBoard.black { blackCount += 1 }
                else if cell == GameBoard.white { whiteCount += 1 }
            }
        }
        return blackCount <= whiteCount ? GameBoard.black : GameBoard.white
    }

    private func checkWinOnBoard(_ board: [[Int]], row: Int, col: Int, player: Int) -> Bool {
        let directions = [(0, 1), (1, 0), (1, 1), (1, -1)]
        let size = GameBoard.size

        for (dr, dc) in directions {
            var count = 1
            for i in 1...4 {
                let r = row + dr * i
                let c = col + dc * i
                guard r >= 0, r < size, c >= 0, c < size, board[r][c] == player else { break }
                count += 1
            }
            for i in 1...4 {
                let r = row - dr * i
                let c = col - dc * i
                guard r >= 0, r < size, c >= 0, c < size, board[r][c] == player else { break }
                count += 1
            }
            if count >= 5 { return true }
        }
        return false
    }

    // MARK: - Evaluation

    func evaluateBoard(board: [[Int]], aiPlayer: Int) -> Int {
        let humanPlayer = aiPlayer == GameBoard.black ? GameBoard.white : GameBoard.black
        var score = 0

        for row in 0..<GameBoard.size {
            for col in 0..<GameBoard.size {
                if board[row][col] == aiPlayer {
                    score += scorePosition(board: board, row: row, col: col, player: aiPlayer)
                } else if board[row][col] == humanPlayer {
                    score -= scorePosition(board: board, row: row, col: col, player: humanPlayer)
                }
            }
        }

        return score
    }

    func scorePosition(board: [[Int]], row: Int, col: Int, player: Int) -> Int {
        var score = 0
        let directions = [(0, 1), (1, 0), (1, 1), (1, -1)]

        for (dr, dc) in directions {
            let (count, blocked) = scoreLine(board: board, row: row, col: col, dr: dr, dc: dc, player: player)
            score += patternScore(count: count, blocked: blocked)
        }

        return score
    }

    func scoreLine(board: [[Int]], row: Int, col: Int, dr: Int, dc: Int, player: Int) -> (count: Int, blocked: Int) {
        var count = 1
        var blocked = 0
        let size = GameBoard.size

        // Forward
        for i in 1...4 {
            let r = row + dr * i
            let c = col + dc * i
            guard r >= 0, r < size, c >= 0, c < size else { blocked += 1; break }
            if board[r][c] == player {
                count += 1
            } else if board[r][c] != GameBoard.empty {
                blocked += 1; break
            } else { break }
        }

        // Backward
        for i in 1...4 {
            let r = row - dr * i
            let c = col - dc * i
            guard r >= 0, r < size, c >= 0, c < size else { blocked += 1; break }
            if board[r][c] == player {
                count += 1
            } else if board[r][c] != GameBoard.empty {
                blocked += 1; break
            } else { break }
        }

        return (count, blocked)
    }

    func patternScore(count: Int, blocked: Int) -> Int {
        switch (count, blocked) {
        case (5, _): return GomokuAI.FIVE
        case (4, 0): return GomokuAI.LIVE_FOUR
        case (4, 1): return GomokuAI.RUSH_FOUR
        case (3, 0): return GomokuAI.LIVE_THREE
        case (3, 1): return GomokuAI.SLEEP_THREE
        case (2, 0): return GomokuAI.LIVE_TWO
        case (2, 1): return GomokuAI.SLEEP_TWO
        case (1, 0): return 10
        case (1, 1): return 1
        default: return 0
        }
    }
}

File: miniprogram/game/GomokuAI.js
import { SIZE, EMPTY, BLACK, WHITE } from './GameBoard.js'

const SCORE_FIVE = 1000000
const SCORE_LIVE_FOUR = 50000
const SCORE_RUSH_FOUR = 5000
const SCORE_LIVE_THREE = 5000
const SCORE_SLEEP_THREE = 500
const SCORE_LIVE_TWO = 500
const SCORE_SLEEP_TWO = 50

export class GomokuAI {
  constructor(depth, candidateRange, maxCandidates, randomPick) {
    this.depth = depth
    this.candidateRange = candidateRange
    this.maxCandidates = maxCandidates
    this.randomPick = randomPick
  }

  findBestMove(board) {
    const local = board.map(r => [...r])
    let candidates = this.getCandidates(local)
    if (candidates.length === 0) return { row: 7, col: 7 }

    if (this.maxCandidates > 0 && candidates.length > this.maxCandidates) {
      candidates = this.selectTopCandidates(local, candidates, this.maxCandidates)
    }

    const scored = []
    for (const move of candidates) {
      local[move.row][move.col] = WHITE
      const score = this.minimax(local, this.depth - 1, -Infinity, Infinity, false, 15)
      local[move.row][move.col] = EMPTY
      scored.push({ move, score })
    }

    scored.sort((a, b) => b.score - a.score)

    if (this.randomPick && scored.length > 1) {
      const top = Math.min(3, scored.length)
      return scored[Math.floor(Math.random() * top)].move
    }
    return scored[0].move
  }

  selectTopCandidates(board, candidates, topN) {
    const scored = candidates.map(move => {
      board[move.row][move.col] = WHITE
      const score = this.quickScore(board, move.row, move.col)
      board[move.row][move.col] = EMPTY
      return { move, score, defense: 0 }
    })

    for (const move of candidates) {
      if (board[move.row][move.col] !== EMPTY) continue
      board[move.row][move.col] = BLACK
      const threat = this.quickScore(board, move.row, move.col)
      board[move.row][move.col] = EMPTY
      const item = scored.find(s => s.move.row === move.row && s.move.col === move.col)
      if (item) item.defense = threat
    }

    scored.sort((a, b) => (b.score + b.defense) - (a.score + a.defense))
    return scored.slice(0, topN).map(s => s.move)
  }

  quickScore(board, row, col) {
    let score = 0
    const dirs = [[0, 1], [1, 0], [1, 1], [1, -1]]
    for (const [dr, dc] of dirs) {
      const count = this.countLine(board, row, col, dr, dc)
      const open = this.countOpenEnds(board, row, col, dr, dc)
      score += this.patternScore(count, open)
    }
    return score
  }

  countLine(board, row, col, dr, dc) {
    const player = board[row][col]
    let count = 1
    for (let i = 1; i < 5; i++) {
      const r = row + dr * i, c = col + dc * i
      if (r >= 0 && r < SIZE && c >= 0 && c < SIZE && board[r][c] === player) count++
      else break
    }
    for (let i = 1; i < 5; i++) {
      const r = row - dr * i, c = col - dc * i
      if (r >= 0 && r < SIZE && c >= 0 && c < SIZE && board[r][c] === player) count++
      else break
    }
    return count
  }

  countOpenEnds(board, row, col, dr, dc) {
    const player = board[row][col]
    let open = 0
    let r = row + dr, c = col + dc
    while (r >= 0 && r < SIZE && c >= 0 && c < SIZE && board[r][c] === player) { r += dr; c += dc }
    if (r >= 0 && r < SIZE && c >= 0 && c < SIZE && board[r][c] === EMPTY) open++
    r = row - dr; c = col - dc
    while (r >= 0 && r < SIZE && c >= 0 && c < SIZE && board[r][c] === player) { r -= dr; c -= dc }
    if (r >= 0 && r < SIZE && c >= 0 && c < SIZE && board[r][c] === EMPTY) open++
    return open
  }

  minimax(board, depth, alpha, beta, maximizing, limit) {
    if (depth === 0) return this.evaluateBoard(board)
    let candidates = this.getCandidates(board)
    if (candidates.length === 0) return this.evaluateBoard(board)
    if (candidates.length > limit) candidates = this.selectTopCandidates(board, candidates, limit)

    if (maximizing) {
      let maxEval = -Infinity
      for (const move of candidates) {
        board[move.row][move.col] = WHITE
        const eval_ = this.minimax(board, depth - 1, alpha, beta, false, limit)
        board[move.row][move.col] = EMPTY
        maxEval = Math.max(maxEval, eval_)
        alpha = Math.max(alpha, eval_)
        if (beta <= alpha) break
      }
      return maxEval
    } else {
      let minEval = Infinity
      for (const move of candidates) {
        board[move.row][move.col] = BLACK
        const eval_ = this.minimax(board, depth - 1, alpha, beta, true, limit)
        board[move.row][move.col] = EMPTY
        minEval = Math.min(minEval, eval_)
        beta = Math.min(beta, eval_)
        if (beta <= alpha) break
      }
      return minEval
    }
  }

  getCandidates(board) {
    const near = Array.from({ length: SIZE }, () => Array(SIZE).fill(false))
    let hasStones = false
    for (let r = 0; r < SIZE; r++) {
      for (let c = 0; c < SIZE; c++) {
        if (board[r][c] !== EMPTY) {
          hasStones = true
          for (let dr = -this.candidateRange; dr <= this.candidateRange; dr++) {
            for (let dc = -this.candidateRange; dc <= this.candidateRange; dc++) {
              const nr = r + dr, nc = c + dc
              if (nr >= 0 && nr < SIZE && nc >= 0 && nc < SIZE && board[nr][nc] === EMPTY) {
                near[nr][nc] = true
              }
            }
          }
        }
      }
    }
    if (!hasStones) return [{ row: 7, col: 7 }]
    const result = []
    for (let r = 0; r < SIZE; r++) {
      for (let c = 0; c < SIZE; c++) {
        if (near[r][c]) result.push({ row: r, col: c })
      }
    }
    return result
  }

  evaluateBoard(board) {
    let aiScore = 0, humanScore = 0
    for (let r = 0; r < SIZE; r++) {
      const line = board[r]
      const [a, h] = this.scoreLine(line)
      aiScore += a; humanScore += h
    }
    for (let c = 0; c < SIZE; c++) {
      const line = Array.from({ length: SIZE }, (_, r) => board[r][c])
      const [a, h] = this.scoreLine(line)
      aiScore += a; humanScore += h
    }
    for (let start = -SIZE + 1; start < SIZE; start++) {
      const line = []
      for (let r = 0; r < SIZE; r++) { const c = r - start; if (c >= 0 && c < SIZE) line.push(board[r][c]) }
      if (line.length >= 2) { const [a, h] = this.scoreLine(line); aiScore += a; humanScore += h }
    }
    for (let start = 0; start < 2 * SIZE - 1; start++) {
      const line = []
      for (let r = 0; r < SIZE; r++) { const c = start - r; if (c >= 0 && c < SIZE) line.push(board[r][c]) }
      if (line.length >= 2) { const [a, h] = this.scoreLine(line); aiScore += a; humanScore += h }
    }
    return aiScore - humanScore
  }

  scoreLine(line) {
    let aiScore = 0, humanScore = 0
    let i = 0
    while (i < line.length) {
      if (line[i] === EMPTY) { i++; continue }
      const player = line[i]
      let count = 1, j = i + 1
      while (j < line.length && line[j] === player) { count++; j++ }
      let openEnds = 0
      if (i > 0 && line[i - 1] === EMPTY) openEnds++
      if (j < line.length && line[j] === EMPTY) openEnds++
      const score = this.patternScore(count, openEnds)
      if (player === WHITE) aiScore += score; else humanScore += score
      i = j
    }
    return [aiScore, humanScore]
  }

  patternScore(count, openEnds) {
    if (count >= 5) return SCORE_FIVE
    if (openEnds === 0) return 0
    if (count === 4) return openEnds === 2 ? SCORE_LIVE_FOUR : SCORE_RUSH_FOUR
    if (count === 3) return openEnds === 2 ? SCORE_LIVE_THREE : SCORE_SLEEP_THREE
    if (count === 2) return openEnds === 2 ? SCORE_LIVE_TWO : SCORE_SLEEP_TWO
    return 0
  }
}
